File: src/kcci/enrich.py
"""Enrich book metadata from OpenLibrary."""

import json
import re
import time
from typing import Optional
import requests
import sqlite3

from . import db

# User agent for API requests
USER_AGENT = "KCCI/1.0 (https://github.com/brianm/kcci; brianm@skife.org)"

# Default delay between API calls
DEFAULT_DELAY = 0.25


def _make_request(url: str, params: dict = None, max_retries: int = 5) -> Optional[requests.Response]:
    """Make HTTP request with exponential backoff on 429 errors."""
    headers = {"User-Agent": USER_AGENT}
    delay = 1.0  # Initial backoff delay

    for attempt in range(max_retries):
        try:
            r = requests.get(url, params=params, headers=headers, timeout=10)

            if r.status_code == 429:
                # Rate limited - exponential backoff
                retry_after = r.headers.get("Retry-After")
                if retry_after:
                    delay = float(retry_after)
                time.sleep(delay)
                delay *= 2  # Exponential backoff
                continue

            r.raise_for_status()
            return r

        except requests.RequestException:
            if attempt < max_retries - 1:
                time.sleep(delay)
                delay *= 2
            else:
                return None

    return None


def normalize_author_for_search(author: str) -> str:
    """Convert 'Last, First' to 'First Last' for API search."""
    if ',' in author:
        parts = author.split(',', 1)
        return f"{parts[1].strip()} {parts[0].strip()}"
    return author.strip()


def normalize_title_for_search(title: str) -> str:
    """Clean up title for API search."""
    # Remove series info in parens
    title = re.sub(r'\s*\([^)]*\)', '', title)
    # Remove subtitle after colon
    title = re.sub(r':.*$', '', title)
    return title.strip()


def search_openlibrary(title: str, authors: list[str]) -> Optional[dict]:
    """Search OpenLibrary for a book by title and author."""
    clean_title = normalize_title_for_search(title)

    # Try with author first
    if authors:
        author = normalize_author_for_search(authors[0])
        params = {"title": clean_title, "author": author, "limit": 5}

        r = _make_request("https://openlibrary.org/search.json", params=params)
        if r:
            data = r.json()
            if data.get("docs"):
                return data["docs"][0]

    # Fall back to title-only search
    r = _make_request(
        "https://openlibrary.org/search.json",
        params={"title": clean_title, "limit": 5},
    )
    if r:
        data = r.json()
        if data.get("docs"):
            return data["docs"][0]

    return None


def get_work_details(work_key: str) -> Optional[dict]:
    """Fetch full work details including description."""
    r = _make_request(f"https://openlibrary.org{work_key}.json")
    if r:
        return r.json()
    return None


def extract_description(work: dict) -> str:
    """Extract description from work details."""
    desc = work.get("description", "")
    if isinstance(desc, dict):
        desc = desc.get("value", "")
    return desc


def enrich_book(conn: sqlite3.Connection, asin: str, title: str, authors: list[str],
                delay: float = DEFAULT_DELAY) -> bool:
    """Enrich a single book with OpenLibrary metadata. Returns True if successful."""
    # Search for the book
    search_result = search_openlibrary(title, authors)
    if not search_result:
        # Save empty metadata to mark as attempted
        db.save_metadata(conn, asin, "", "", [], None, None)
        return False

    work_key = search_result.get("key", "")
    subjects = search_result.get("subject", [])[:20]  # Limit subjects
    publish_year = search_result.get("first_publish_year")
    isbn = None
    if search_result.get("isbn"):
        isbn = search_result["isbn"][0]

    # Get full work details for description
    description = ""
    if work_key:
        time.sleep(delay)  # Rate limit between API calls
        work = get_work_details(work_key)
        if work:
            description = extract_description(work)

    db.save_metadata(conn, asin, work_key, description, subjects, isbn, publish_year)
    return bool(description)


def enrich_batch(conn: sqlite3.Connection, limit: Optional[int] = None, delay: float = DEFAULT_DELAY,
                 progress_callback=None) -> tuple[int, int]:
    """Enrich a batch of books. Returns (success_count, total_attempted)."""
    books = db.get_books_without_metadata(conn, limit)
    success = 0

    for i, book in enumerate(books):
        authors = json.loads(book["authors"])
        if enrich_book(conn, book["asin"], book["title"], authors, delay):
            success += 1

        if progress_callback:
            progress_callback(i + 1, len(books), book["title"])

        if i < len(books) - 1:
            time.sleep(delay)  # Rate limit between books

    return success, len(books)

File: src-tauri/src/enrich.rs
use regex::Regex;
use reqwest::blocking::Client;
use std::sync::OnceLock;
use std::thread;
use std::time::Duration;

use crate::db::EnrichmentData;
use crate::error::Result;

const USER_AGENT: &str = "Ook/1.0 (https://github.com/brianm/ook; brianm@skife.org)";
const DEFAULT_DELAY: Duration = Duration::from_millis(250);

// Static regexes for title normalization (compiled once)
static PARENTHETICAL_RE: OnceLock<Regex> = OnceLock::new();
static SUBTITLE_RE: OnceLock<Regex> = OnceLock::new();

fn get_parenthetical_regex() -> &'static Regex {
    PARENTHETICAL_RE.get_or_init(|| Regex::new(r"\s*\([^)]*\)").expect("Invalid parenthetical regex"))
}

fn get_subtitle_regex() -> &'static Regex {
    SUBTITLE_RE.get_or_init(|| Regex::new(r":.*$").expect("Invalid subtitle regex"))
}

/// OpenLibrary API client with rate limiting
pub struct OpenLibrary {
    client: Client,
}

impl OpenLibrary {
    /// Create a new OpenLibrary client
    pub fn new() -> Result<Self> {
        let client = Client::builder()
            .user_agent(USER_AGENT)
            .timeout(Duration::from_secs(10))
            .build()?;
        Ok(Self { client })
    }

    /// Search for a book and fetch its metadata
    pub fn search(&self, title: &str, authors: &[String]) -> Result<Option<EnrichmentData>> {
        let clean_title = normalize_title(title);

        // Try with author first
        if !authors.is_empty() {
            let author = normalize_author(&authors[0]);
            if let Some(result) = self.search_api(&clean_title, Some(&author))? {
                return Ok(Some(result));
            }
        }

        // Fallback to title-only search
        self.search_api(&clean_title, None)
    }

    /// Perform the actual API search
    fn search_api(
        &self,
        title: &str,
        author: Option<&str>,
    ) -> Result<Option<EnrichmentData>> {
        let mut url = format!(
            "https://openlibrary.org/search.json?title={}&limit=5&fields=key,title,author_name,subject,isbn,first_publish_year",
            urlencoding::encode(title)
        );
        if let Some(a) = author {
            url.push_str(&format!("&author={}", urlencoding::encode(a)));
        }

        let Some(response) = self.request_with_backoff(&url)? else {
            return Ok(None);
        };

        let data: serde_json::Value = response.json()?;
        let docs = data.get("docs").and_then(|d| d.as_array());

        if let Some(docs) = docs {
            if let Some(first) = docs.first() {
                let work_key = first.get("key").and_then(|k| k.as_str()).unwrap_or("");

                let subjects: Vec<String> = first
                    .get("subject")
                    .and_then(|s| s.as_array())
                    .map(|arr| {
                        arr.iter()
                            .filter_map(|v| v.as_str())
                            .take(20)
                            .map(|s| s.to_string())
                            .collect()
                    })
                    .unwrap_or_default();

                let isbn = first
                    .get("isbn")
                    .and_then(|i| i.as_array())
                    .and_then(|arr| arr.first())
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());

                let publish_year = first
                    .get("first_publish_year")
                    .and_then(|y| y.as_i64())
                    .map(|y| y as i32);

                // Get description from work details
                thread::sleep(DEFAULT_DELAY);
                let description = if !work_key.is_empty() {
                    self.get_work_description(work_key)?
                } else {
                    String::new()
                };

                return Ok(Some(EnrichmentData {
                    openlibrary_key: work_key.to_string(),
                    description,
                    subjects,
                    isbn,
                    publish_year,
                }));
            }
        }

        Ok(None)
    }

    /// Fetch work description from OpenLibrary
    fn get_work_description(&self, work_key: &str) -> Result<String> {
        let url = format!("https://openlibrary.org{}.json", work_key);
        if let Some(resp) = self.request_with_backoff(&url)? {
            let data: serde_json::Value = resp.json()?;
            if let Some(desc) = data.get("description") {
                if let Some(s) = desc.as_str() {
                    return Ok(s.to_string());
                }
                if let Some(obj) = desc.as_object() {
                    if let Some(v) = obj.get("value").and_then(|v| v.as_str()) {
                        return Ok(v.to_string());
                    }
                }
            }
        }
        Ok(String::new())
    }

    /// Make HTTP request with exponential backoff on 429 errors
    fn request_with_backoff(
        &self,
        url: &str,
    ) -> Result<Option<reqwest::blocking::Response>> {
        let mut delay = Duration::from_secs(1);
        let max_retries = 5;

        for attempt in 0..max_retries {
            match self.client.get(url).send() {
                Ok(resp) if resp.status() == 429 => {
                    // Rate limited - check Retry-After header
                    if let Some(retry) = resp.headers().get("Retry-After") {
                        if let Ok(secs) = retry.to_str().unwrap_or("1").parse::<u64>() {
                            delay = Duration::from_secs(secs);
                        }
                    }
                    thread::sleep(delay);
                    delay *= 2;
                }
                Ok(resp) if resp.status().is_success() => return Ok(Some(resp)),
                Ok(_) => return Ok(None),
                Err(_) if attempt < max_retries - 1 => {
                    thread::sleep(delay);
                    delay *= 2;
                }
                Err(e) => return Err(e.into()),
            }
        }
        Ok(None)
    }
}

/// Normalize title for API search (remove series info, subtitles)
fn normalize_title(title: &str) -> String {
    let cleaned = get_parenthetical_regex().replace_all(title, "");
    get_subtitle_regex().replace_all(&cleaned, "").trim().to_string()
}

/// Normalize author name (convert "Last, First" to "First Last")
fn normalize_author(author: &str) -> String {
    if let Some((last, first)) = author.split_once(',') {
        format!("{} {}", first.trim(), last.trim())
    } else {
        author.trim().to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_title() {
        assert_eq!(normalize_title("The Book (Series #1)"), "The Book");
        assert_eq!(normalize_title("Title: A Subtitle"), "Title");
        assert_eq!(
            normalize_title("Book (Edition 2): More Stuff"),
            "Book"
        );
    }

    #[test]
    fn test_normalize_author() {
        assert_eq!(normalize_author("Doe, John"), "John Doe");
        assert_eq!(normalize_author("Jane Smith"), "Jane Smith");
    }
}
